test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/fields/ObjectFieldDataGenerator.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.logsdb.datageneration.fields;
+
+import org.elasticsearch.core.CheckedConsumer;
+import org.elasticsearch.logsdb.datageneration.DataGeneratorSpecification;
+import org.elasticsearch.logsdb.datageneration.FieldDataGenerator;
+import org.elasticsearch.logsdb.datageneration.FieldType;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ObjectFieldDataGenerator implements FieldDataGenerator {
+    private final DataGeneratorSpecification specification;
+    private final int depth;
+
+    private final List<ChildField> childFields;
```

Comment:
Why not a list of `FieldDataGenerator`...that would allow us to create aribitrary nested hierachies right? Each element could then be an ObjectFieldDataGenerator or a leaf type...BTW, maybe for leaf types it might make sense to include a base class `LeafFieldDataGenerator` parent of types like `KeywordFieldDataGenerator`,...whether nested fields are supported otr not might be controlled by mens of `subobjects` in `ObjectFieldDataGenerator`.

Reply:
A `ChildField` contains `FieldDataGenerator` inside of it and we do support nested hierarchies as you say. `ChildField` holds a generator and field name, fields are not responsible for writing their names to mapping and documents. It is primarily done this way because top level object generator does not have a name and if fields were responsible for writing names themselves we have to deal with that edge case.

Reply:
See `DataGeneratorSnapshotTests` which contains an example with nested object hierarchy.